File: internal/http/git_status.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package http

import (
	"net/http"
)

// handleGetGitStatus handles GET /api/p/{project}/git/status.
// Returns a git status summary for the project, or {"available":false}
// when the project directory is not a git repository.
func (s *Server) handleGetGitStatus(w http.ResponseWriter, r *http.Request) {
	p := projectFromCtx(r.Context())
	if p == nil {
		writeJSON(w, http.StatusInternalServerError, apiError("no_project", "no project in context"))
		return
	}

	if p.Git == nil {
		writeJSON(w, http.StatusOK, map[string]any{"available": false})
		return
	}

	summary, err := p.Git.Status()
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, apiError("git_error", err.Error()))
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"available":    true,
		"branch":       summary.Branch,
		"dirty":        summary.Dirty,
		"head_sha":     summary.HeadSHA,
		"head_message": summary.HeadMessage,
		"head_author":  summary.HeadAuthor,
		"head_when":    summary.HeadWhen,
	})
}

File: web/src/stores/gitStatus.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { ref } from 'vue'
import { defineStore } from 'pinia'
import { fetchGitStatus } from '@/api/git'
import type { GitStatusResponse } from '@/types/api'

export const useGitStatusStore = defineStore('gitStatus', () => {
  const available = ref(false)
  const branch = ref('')
  const dirty = ref(false)
  const headSha = ref('')
  const headMessage = ref('')
  const headAuthor = ref('')
  const headWhen = ref('')

  function reset() {
    available.value = false
    branch.value = ''
    dirty.value = false
    headSha.value = ''
    headMessage.value = ''
    headAuthor.value = ''
    headWhen.value = ''
  }

  async function fetch(project: string) {
    reset()
    try {
      const res = await fetchGitStatus(project)
      applyWsEvent(res)
    } catch {
      available.value = false
    }
  }

  function applyWsEvent(data: GitStatusResponse) {
    available.value = data.available
    branch.value = data.branch ?? ''
    dirty.value = data.dirty ?? false
    headSha.value = data.head_sha ?? ''
    headMessage.value = data.head_message ?? ''
    headAuthor.value = data.head_author ?? ''
    headWhen.value = data.head_when ?? ''
  }

  return { available, branch, dirty, headSha, headMessage, headAuthor, headWhen, fetch, applyWsEvent }
})
